src/main/java/dp/handler/Handler.java

```diff
@@ -353,11 +353,17 @@ private String getDownloadUrl(boolean isPublished, String filePath, WorkbookDeta
         return details.getDownloadURI();
     }
 
-    public String getS3URL(String url) {
-        if (StringUtils.isEmpty(bucketUrl) || StringUtils.isEmpty(bucketS3Url)) {
-            return url;
+    /** 
+     * getS3URL returns an S3-compatible version of url:
+     * replaces the CloudFront/S3 bucket URL (bucketUrl) prefix in url
+     * with the more S3-lib-friendly prefix (bucketS3Url).
+     * Both vars must be non-empty for the replacement to occur.
+     */
+    String getS3URL(String url) {
```

Comment:
Do we not need to whether the method is private or public (or protected etc.)

Reply:
I was going to change this so it looked like the following:

```
/** 
     * s3URL returns an S3-compatible version of url:
     * replaces the CloudFront/S3 bucket URL (bucketUrl) prefix in url
     * with the more S3-lib-friendly prefix (bucketS3Url).
     * Both vars must be non-empty for the replacement to occur.
     */
    private String s3URL(String url) {
        if (StringUtils.isNotEmpty(bucketUrl) && StringUtils.isNotEmpty(bucketS3Url)) {
            url = url.replace(bucketUrl, bucketS3Url);
        }
        return url;
    }

    public String getS3URL(String url) {
        return s3URL(url)
    }
```

So tests still pass and the logic is hidden in private method?

Reply:
There are 4 access modifiers in Java: `public`, `private`, `protected` & `default` (none).

No modifier means _package private_ which is the same as _private_ in Go i.e. only objects defined within the same package can access it.

Reply:
Thanks for the explanation, then what Gedge has done does look superior to my solution.